jerry-core/vm/opcodes.c

```diff
@@ -575,43 +575,54 @@ opfunc_append_array (ecma_value_t *stack_top_p, /**< current stack top */
  *
  * @return executable object
  */
-ecma_value_t
-opfunc_create_executable_object (vm_frame_ctx_t *frame_ctx_p) /**< frame context */
+vm_executable_object_t *
+opfunc_create_executable_object (vm_frame_ctx_t *frame_ctx_p, /**< frame context */
+                                 create_executable_object_type_t type) /**< executable object type */
 {
   const ecma_compiled_code_t *bytecode_header_p = frame_ctx_p->bytecode_header_p;
-  size_t size;
+  size_t size, register_end;
 
   ecma_bytecode_ref ((ecma_compiled_code_t *) bytecode_header_p);
 
   if (bytecode_header_p->status_flags & CBC_CODE_FLAGS_UINT16_ARGUMENTS)
   {
     cbc_uint16_arguments_t *args_p = (cbc_uint16_arguments_t *) bytecode_header_p;
-    size = ((size_t) args_p->register_end + (size_t) args_p->stack_limit) * sizeof (ecma_value_t);
+    register_end = (size_t) args_p->register_end;
+    size = (register_end + (size_t) args_p->stack_limit) * sizeof (ecma_value_t);
   }
   else
   {
     cbc_uint8_arguments_t *args_p = (cbc_uint8_arguments_t *) bytecode_header_p;
-    size = ((size_t) args_p->register_end + (size_t) args_p->stack_limit) * sizeof (ecma_value_t);
+    register_end = (size_t) args_p->register_end;
+    size = (register_end + (size_t) args_p->stack_limit) * sizeof (ecma_value_t);
   }
 
   size_t total_size = JERRY_ALIGNUP (sizeof (vm_executable_object_t) + size, sizeof (uintptr_t));
 
-  ecma_object_t *proto_p = ecma_op_get_prototype_from_constructor (JERRY_CONTEXT (current_function_obj_p),
-                                                                   ECMA_BUILTIN_ID_GENERATOR_PROTOTYPE);
+  ecma_object_t *proto_p = NULL;
+
+  if (type == CREATE_EXECUTABLE_OBJECT_GENERATOR)
+  {
+    proto_p = ecma_op_get_prototype_from_constructor (JERRY_CONTEXT (current_function_obj_p),
+                                                      ECMA_BUILTIN_ID_GENERATOR_PROTOTYPE);
+  }
 
   ecma_object_t *object_p = ecma_create_object (proto_p,
                                                 total_size,
                                                 ECMA_OBJECT_TYPE_CLASS);
 
-  ecma_deref_object (proto_p);
-
   vm_executable_object_t *executable_object_p = (vm_executable_object_t *) object_p;
 
+  if (type == CREATE_EXECUTABLE_OBJECT_GENERATOR)
+  {
+    ecma_deref_object (proto_p);
+  }
+
+  /* Async function objects are not accessible, so their class_id is not relevant. */
```

Comment:
That's definitely not correct:

```js
async function f () {}
Object.prototype.toString.call(f)
>> "[object AsyncFunction]"
```

Reply:
That is a different thing. Currently the prototype objects defined by the standard (e.g. AsyncFunction) has not been implemented yet , this patch focuses on the core things.

When 'f' is called, it creates another hidden object with the execution state, and this object is never visible. It does not even need to be an object, but gc likes this. It could be dropped to reduce memory consumption at some point.

Reply:
Thanks for the explanation!